src/app/qgsmaptoolscalefeature.cpp

```diff
@@ -0,0 +1,479 @@
+  mMagnetSpinBox->setClearValue( 0, tr( "No snapping" ) );
+  //mMagnetSpinBox->setSizePolicy( QSizePolicy::MinimumExpanding, QSizePolicy::Preferred );
+  mMagnetSpinBox->setMinimumWidth( 120 );
+  mLayout->addWidget( mMagnetSpinBox );
+
+  // connect signals
+  mScaleSpinBox->installEventFilter( this );
+  connect( mScaleSpinBox, static_cast < void ( QgsDoubleSpinBox::* )( double ) > ( &QgsDoubleSpinBox::valueChanged ), this, &QgsScaleMagnetWidget::scaleSpinBoxValueChanged );
+
+  // config focus
+  setFocusProxy( mScaleSpinBox );
+}
+
+void QgsScaleMagnetWidget::setScale( double scale )
+{
+  const int m = magnet();
+  if ( m )
+  {
+    mScaleSpinBox->setValue( std::round( scale / m ) * m );
+  }
+  else
+  {
+    mScaleSpinBox->setValue( scale );
+  }
+}
+
+double QgsScaleMagnetWidget::scale() const
+{
+  return mScaleSpinBox->value();
+}
+
+void QgsScaleMagnetWidget::setMagnet( int magnet )
+{
+  mMagnetSpinBox->setValue( magnet );
+}
+
+int QgsScaleMagnetWidget::magnet() const
+{
+  return mMagnetSpinBox->value();
+}
+
+
+bool QgsScaleMagnetWidget::eventFilter( QObject *obj, QEvent *ev )
+{
+  if ( obj == mScaleSpinBox && ev->type() == QEvent::KeyPress )
+  {
+    QKeyEvent *event = static_cast<QKeyEvent *>( ev );
+    if ( event->key() == Qt::Key_Escape )
+    {
+      emit scaleEditingCanceled();
+      return true;
+    }
+    if ( event->key() == Qt::Key_Enter || event->key() == Qt::Key_Return )
+    {
+      emit scaleEditingFinished( scale() );
+      return true;
+    }
+  }
+
+  return false;
+}
+
+void QgsScaleMagnetWidget::scaleSpinBoxValueChanged( double scale )
+{
+  emit scaleChanged( scale );
+}
+
+QgsMapToolScaleFeature::QgsMapToolScaleFeature( QgsMapCanvas *canvas )
+  : QgsMapToolEdit( canvas )
+  , mScaling( 0 )
+  , mScalingActive( false )
+{
+}
+
+QgsMapToolScaleFeature::~QgsMapToolScaleFeature()
+{
+  deleteScalingWidget();
+  mAnchorPoint.reset();
+  deleteRubberband();
+}
+
+void QgsMapToolScaleFeature::canvasMoveEvent( QgsMapMouseEvent *e )
+{
+  if ( mScalingActive )
+  {
+    const double distance = mFeatureCenter.distance( toLayerCoordinates( mLayer, e->mapPoint() ) );
+    double scale =  distance / mBaseDistance; // min 0 or no limit?
+
+    if ( mScalingWidget )
+    {
+      disconnect( mScalingWidget, &QgsScaleMagnetWidget::scaleChanged, this, &QgsMapToolScaleFeature::updateRubberband );
+      mScalingWidget->setScale( scale );
+      mScalingWidget->setFocus( Qt::TabFocusReason );
+      mScalingWidget->editor()->selectAll();
+      connect( mScalingWidget, &QgsScaleMagnetWidget::scaleChanged, this, &QgsMapToolScaleFeature::updateRubberband );
+      if ( mScalingWidget->magnet() )
+      {
+        scale = mScalingWidget->scale();
+      }
+    }
+    updateRubberband( scale );
+  }
+}
+
+
+void QgsMapToolScaleFeature::canvasReleaseEvent( QgsMapMouseEvent *e )
+{
+  if ( !mCanvas )
+  {
+    return;
+  }
+
+  //QgsVectorLayer *mLayer = currentVectorLayer();
+  if ( !mLayer )
```

Comment:
What's the rationale behind the change to storing the layer instead of using currentVectorLayer()?

Reply:
Easier retrieval mostly.